src/VisualStudio/Core/Def/Implementation/AbstractEditorFactory.cs

```diff
@@ -252,54 +253,49 @@ int IVsEditorFactoryNotify.NotifyItemRenamed(IVsHierarchy pHier, uint itemid, st
+                    var documentId = DocumentId.CreateNewId(projectId);
+                    var forkedSolution = solution.AddDocument(DocumentInfo.Create(documentId, filePath, loader: new FileTextLoader(filePath, defaultEncoding: null), filePath: filePath));
+                    var addedDocument = forkedSolution.GetDocument(documentId);
+
+                    var rootToFormat = addedDocument.GetSyntaxRootAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+
+                    var formattedTextChanges = Formatter.GetFormattedTextChanges(rootToFormat, workspace, addedDocument.Options, cancellationToken);
+                    var formattedText = addedDocument.GetTextAsync(cancellationToken).WaitAndGetResult(cancellationToken).WithChanges(formattedTextChanges);
+
+                    // Ensure the line endings are normalized. The formatter doesn't touch everything if it doesn't need to.
+                    string targetLineEnding = addedDocument.Options.GetOption(FormattingOptions.NewLine);
+
+                    var originalText = formattedText;
+                    foreach (var originalLine in originalText.Lines)
                     {
-                        cancellationToken.ThrowIfCancellationRequested();
+                        string originalNewLine = originalText.ToString(CodeAnalysis.Text.TextSpan.FromBounds(originalLine.End, originalLine.EndIncludingLineBreak));
 
-                        // If there is a linebreak at all...
-                        if (line.LineBreakLength != 0)
+                        // Check if we have a line ending, so we don't go adding one to the end if we don't need to.
+                        if (originalNewLine.Length > 0 && originalNewLine != targetLineEnding)
                         {
-                            if (line.GetLineBreakText() != "\r\n")
-                            {
-                                edit.Replace(line.End, line.LineBreakLength, "\r\n");
-                            }
+                            var currentLine = formattedText.Lines[originalLine.LineNumber];
+                            var currentSpan = CodeAnalysis.Text.TextSpan.FromBounds(currentLine.End, currentLine.EndIncludingLineBreak);
+                            formattedText = formattedText.WithChanges(new TextChange(currentSpan, targetLineEnding));
                         }
                     }
 
-                    edit.Apply();
-                }
-
-                var formattedTextChanges = GetFormattedTextChanges(ComponentModel.GetService<VisualStudioWorkspace>(), filePath, textDocument.TextBuffer.CurrentSnapshot.AsText(), cancellationToken);
-
-                using (var edit = textDocument.TextBuffer.CreateEdit())
-                {
-                    foreach (var change in formattedTextChanges)
+                    using (var textWriter = new StreamWriter(filePath, append: false, encoding: formattedText.Encoding))
```

Comment:
I have a RepeatIOAsync helper inside the symbol search stuff.  Perhaps it could be repurposed.  I think, in general, retrying a few times (while we swallow a few known exception types) is fine.  Otherwise we have the simple problem where we can throw just because windows defender locked the file at an inopportune time.